In [83]:
!pip3 install --user okpy

from lecture import *

You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


# Introduction to programming in Python

# Lecture 2

Learning objectives:

* Know how to modify elements in a list.
* Be able iterate through different combinations of lists.
* Know how to use a *tuple* to store data elements and understand how it differs from a *list*.
* Know the difference between a locally-scoped and globally-scoped variables.
* Be able to use an *if* statement to execute some code blocks conditionally.
* Learn how to compute using [Numerical Python (*NumPy*)](http://www.numpy.org/).
* Know how to handle multidimensional arrays.

# Changing elements in a list
Say we want to add $2$ to all the numbers in a list:

In [38]:
v = [-1, 1, 10]
for e in v:
    e=e+2
print(v)

[-1, 1, 10]


We can see that the list *v* is unaltered! The reason for this is that inside the loop, *e* is an ordinary (int) variable. At the start of the iteration *e* is assigned a *copy* of the next element in the list. Inside the loop we can change *e* but *v* itself is unaltered. If we want to change *v* we have to use an index to access and modify its elements:

In [39]:
v[1] = 4 # assign 4 to 2nd element (index 1) in v
print(v)

[-1, 4, 10]


To add 2 to all values we need a *for* loop over indices:

In [40]:
for i in range(len(v)):
    v[i] = v[i] + 2
print(v)

[1, 6, 12]


## Exercise 2.1: Create a list and modify it.
* Create a list of all integers in the range -10 to 10 (call the list *vector* for testing purposes).
* Write a loop to multiple each element of the list by 2.

In [41]:
a = -11
vector = []
while a <= 9:
    a +=1
    vector.append(a)
for i in range(len(vector)):
    vector[i] = 2 * vector[i]
print(vector)

[-20, -18, -16, -14, -12, -10, -8, -6, -4, -2, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


In [42]:
ok.grade('question-2_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Traversing multiple lists simultaneously - *zip(list1, list2, ...)*
Consider how we can loop over elements in both Cdegrees and Fdegrees at the same time. One approach would be to use list indices:

In [43]:
# First we have to recreate the data from the previous lecture
Cdegrees = [deg for deg in range(-20, 41, 5)]
Fdegrees = [(9/5)*deg + 32 for deg in Cdegrees]

for i in range(len(Cdegrees)):
    print(Cdegrees[i], Fdegrees[i])

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


An alternative construct (regarded as more ”Pythonic”) uses the *zip* function:

In [44]:
for C, F in zip(Cdegrees, Fdegrees):
    print(C, F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


Another example with three lists:

In [45]:
l1 = [3, 6, 1]; l2 = [1.5, 1, 0]; l3 = [9.1, 3, 2]
for e1, e2, e3 in zip(l1, l2, l3):
    print(e1, e2, e3)

3 1.5 9.1
6 1 3
1 0 2


If the lists are of unequal length then the loop stops when the end of the shortest list is reached. Experiment with this:

In [46]:
l1 = [3, 6, 1, 4, 6]; l2 = [1.5, 1, 0, 7]; l3 = [9.1, 3, 2, 0, 9]
for e1, e2, e3 in zip(l1, l2, l3):
    print(e1, e2, e3)

3 1.5 9.1
6 1 3
1 0 2
4 7 0


## Nested lists: list of lists
A *list* can contain **any** object, including another *list*. To illustrate this, consider how to store the conversion table as a single Python list rather than two separate lists.

In [47]:
Cdegrees = range(-20, 41, 5)
Fdegrees = [(9.0/5)*C + 32 for C in Cdegrees]
table1 = [Cdegrees, Fdegrees]  # List of two lists
print("table1 = ", table1)
print("table1[0] = ", table1[0])
print("table1[1] = ", table1[1])
print("table1[1][3] = ", table1[1][3])

table1 =  [range(-20, 41, 5), [-4.0, 5.0, 14.0, 23.0, 32.0, 41.0, 50.0, 59.0, 68.0, 77.0, 86.0, 95.0, 104.0]]
table1[0] =  range(-20, 41, 5)
table1[1] =  [-4.0, 5.0, 14.0, 23.0, 32.0, 41.0, 50.0, 59.0, 68.0, 77.0, 86.0, 95.0, 104.0]
table1[1][3] =  23.0


This gives us a table of two rows. How do we create a table of columns instead:

In [48]:
table2 = []
for C, F in zip(Cdegrees, Fdegrees):
    row = [C, F]
    table2.append(row)
print(table2)

[[-20, -4.0], [-15, 5.0], [-10, 14.0], [-5, 23.0], [0, 32.0], [5, 41.0], [10, 50.0], [15, 59.0], [20, 68.0], [25, 77.0], [30, 86.0], [35, 95.0], [40, 104.0]]


We can use list comprehension to do this more elegantly:

In [49]:
table2 = [[C, F] for C, F in zip(Cdegrees, Fdegrees)]
print(table2)

[[-20, -4.0], [-15, 5.0], [-10, 14.0], [-5, 23.0], [0, 32.0], [5, 41.0], [10, 50.0], [15, 59.0], [20, 68.0], [25, 77.0], [30, 86.0], [35, 95.0], [40, 104.0]]


And you can loop through this list as before:

In [50]:
for C,F in table2:
    print(C,F)

-20 -4.0
-15 5.0
-10 14.0
-5 23.0
0 32.0
5 41.0
10 50.0
15 59.0
20 68.0
25 77.0
30 86.0
35 95.0
40 104.0


## Tuples: lists that cannot be changed
Tuples are **constant** lists, i.e. you can use them in much the same way as lists except you cannot modify them. They are an example of an [**immutable**](http://en.wikipedia.org/wiki/Immutable_object) type.

In [51]:
t = (2, 4, 6, 'temp.pdf')               # Define a tuple.
t =  2, 4, 6, 'temp.pdf'                # Can skip parenthesis as it is assumed in this context.

Let's see what happens when we try to modify the tuple like we did with a list:

```python
t[1] = -1

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-593c03edf054> in <module>()
----> 1 t[1] = -1

TypeError: 'tuple' object does not support item assignment```

```python
t.append(0)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-19-78592bf72d62> in <module>()
----> 1 t.append(0)

AttributeError: 'tuple' object has no attribute 'append'```

```python
del t[1]

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-20-0193a527a912> in <module>()
----> 1 del t[1]

TypeError: 'tuple' object doesn't support item deletion```

However, we can use the tuple to compose a new tuple:

In [52]:
t = t + (-1.0, -2.0)
print(t)

(2, 4, 6, 'temp.pdf', -1.0, -2.0)


In [53]:
a=[1,2,4];b=[2,4,5,6]
c=[a,b]
print(c)


[[1, 2, 4], [2, 4, 5, 6]]


So, why would we use tuples when lists have more functionality?

* Tuples are constant and thus protected against accidental changes.
* Tuples are faster than lists.
* Tuples are widely used in Python software (so you need to know about tuples to understand other people's code!)
* Tuples (but not lists) can be used as keys in dictionaries (more about dictionaries later).

## Exercise 2.2: Make a table (a list of lists) of function values
* Write a loop that evaluates the expression $y(t) = v_0 t − 0.5gt^2$ for 11 evenly spaced values ranging from 0, to $2v_0/g$ (remember that dividing a range into n intervals results in n+1 values!) You can assume that $v_0 = 1$, $g=9.81 ms^{-2}$.
* Store the time values and displacement ($y$) values as a nested list, i.e.
```python
tlist = [t0, t1, t2, ...]
ylist = [y0, y1, y2, ...]
displacement = [tlist, ylist]
```
* Use the variable names tlist, ylist and displacement as illustrated above example for testing purposes.

In [54]:
v0=1;g=9.81;a=0;num=11;n=0;tlist=[0];ylist=[]
while n<(2*v0/g):
    d = ((2*v0/g)+a)/(num-1)
    n = n+d
    tlist.append(n)
for i in range(len(tlist)):
    ylist.append(v0*tlist[i]-0.5*g*(tlist[i])**2)
displacement=[tlist,ylist]
print(displacement)
    

[[0, 0.02038735983690112, 0.04077471967380224, 0.061162079510703356, 0.08154943934760447, 0.1019367991845056, 0.12232415902140673, 0.14271151885830785, 0.16309887869520898, 0.1834862385321101, 0.20387359836901123], [0.0, 0.018348623853211007, 0.03261977573904179, 0.04281345565749235, 0.04892966360856269, 0.0509683995922528, 0.048929663608562685, 0.04281345565749235, 0.03261977573904179, 0.018348623853210982, -2.7755575615628914e-17]]


In [55]:
ok.grade('question-2_2')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 9
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 9}

## The *if* construct
Consider this simple example:

In [56]:
from math import sin, pi

def f(x):
    if 0 <= x <= pi:
        return sin(x)
    else:
        return 0
print(f(-pi/2), f(pi/2), f(3*pi/2))

0 1.0 0


Sometimes it is clearer to write this as an *inline* statement:

In [57]:
from math import pi, sin
def f(x):
    return (sin(x) if 0 <= x <= pi else 0)
print(f(-pi/2), f(pi/2), f(3*pi/2))

0 1.0 0


In general (the *else* block can be skipped if there are no statements to be executed when False) we can put together multiple conditions. Only the first condition that is True is executed.

```
if condition1:
    <block of statements, executed if condition1 is True>
elif condition2:
    <block of statements>
elif condition3:
    <block of statements>
else:
    <block of statements>
    
<next statement of the program>
```

## Exercise 2.3: Express a step function as a Python function
The following "step" function is known as the Heaviside function and
is widely used in mathematics:
$$H(x)=\begin{cases}0, & \text{if $x<0$}.\\\\
1, & \text{if $x\ge 0$}.\end{cases}$$
Write a Python function heaviside(x) that computes H(x).

In [58]:
def heaviside(x):
    if x < 0:
        return 0
    elif x >=0:
        return 1
    else:
        return ()
print(heaviside(-1),heaviside(1))



0 1


In [59]:
ok.grade('question-2_3')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 5
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 5}

## Exercise 2.4: Implement the factorial function

The factorial of $n$, written as $n!$, is defined as

$$n! = n(n − 1)(n − 2) \cdots 2 \cdot 1,$$

with the special cases

$$1! = 1,$$ $$0! = 1.$$

For example, $4! = 4 \cdot 3 \cdot 2 \cdot 1 = 24$, and $2! = 2 \cdot 1 = 2$.

Implement your own factorial function to calculate $n!$. Return 1 immediately if $n$ is 1 or 0, otherwise use a loop to compute $n!$. You can use Pythons own [math.factorial(x)](https://docs.python.org/3/library/math.html) to check your code.

In [60]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def my_factorial(n):
    if n ==0:
        return 1
    elif n ==1:
        return 1
    else:
        a = 1
        my_fact = 1
        while n >=2 :
            my_fact *= n
            n = n - 1
        return my_fact           
print(my_factorial(7))

5040


In [61]:
ok.grade('question-2_4')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 5
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 5}

## Exercise 2.5: Compute the length of a path

Some object is moving along a path in the plane. At $n$ points of time we have recorded the corresponding $(x, y)$ positions of the object:
$(x_0, y_0), (x_1, y_1), \ldots, (x_{n-1}, y_{n-1})$. The total length $L$ of the path from $(x_0, y_0)$ to $(x_{n-1}, y_{n-1})$ is the sum of all the individual line segments $(x_{i-1}, y_{i-1})$ to $(x_i, y_i)$, $i = 1, \ldots, n-1$:

$$L = \sum_{i=1}^{n-1}{\sqrt{(x_i - x_{i-1})^2 + (y_i - y_{i-1})^2}}.$$

Create a function *pathlength(x, y)* for computing $L$ according to the formula. The arguments $x$ and $y$ are two lists that hold all the $x_0, \ldots, x_{n-1}$ and $y_0, \ldots, y_{n-1}$ coordinates, respectively. Test the function on a triangular path with the four points (1, 1), (2, 1), (1, 2), and (1, 1).

In [62]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def path_length(x, y):
    n = len(x)
    my_length = 0
    i=1
    while i < n:
        length = ((x[i]-x[i-1])**2 + (y[i]-y[i-1])**2)**0.5
        my_length = my_length + length
        i+=1
    return my_length
print(path_length([1,2,1,1],[1,1,2,1]))
        

3.414213562373095


In [63]:
ok.grade('question-2_5')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 4}

## Exercise 2.6: Approximate $\pi$

As you know the circumference of a circle is given by $2 \pi r$ where $r$ is the radius of the cirlce.  $\pi$ is therefore the circumference of a circle with $r= \frac{1}{2}$.

We can approximate this circumference by a many-sided polygon through points on the circle. The sum of the lengths of the sides of the polygon will approximate the circumference.

Firstly compute $n+1$ points around a circle according to the formulae:

$ x_i = \frac{1}{2} \cos(\frac{2 \pi i}{n}),   y_i = \frac{1}{2} \sin(\frac{2 \pi i}{n}),   i = 0 \cdots n$

Then use your `path_length` function from the previous question to approximate $\pi$.  Name your function for estimating $\pi$ `approx_pi` for testing purposes.

In [64]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def approx_pi(n):
    from math import cos, sin, sqrt, pi
    i = 0
    x = []
    y = []
    a = 1
    while i <= n:
        x.append((1/2)*cos((2*pi*i)/n))
        y.append((1/2)*sin((2*pi*i)/n))
        i+=1
    approx_pi = path_length(x,y)
    return approx_pi
print(approx_pi(4))
        


2.82842712474619


In [65]:
ok.grade('question-2_6')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Exercise 2.7: Make a list of prime numbers

Define a function called `prime_list` that lists all the prime numbers up to a given $n$.  

Save all the prime numbers up to $100$ in a list called `myprimes`.

Hint: google the Sieve of Eratosthenes.


In [66]:
# Uncomment and complete this code - keep the names the same for testing purposes. 

def prime_list(n):
    list1 = []
    for num in range(2,n + 1):
        prime = True
        for i in range(2,num):
            if (num%i) == 0:
                prime = False
        if prime:
            list1.append(num)
    return list1
myprimes = prime_list(100)
print(myprimes)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


In [67]:
ok.grade('question-2_7')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Vectors and arrays

You have known **vectors** since high school mathematics, *e.g.*, point $(x,y)$ in the plane, point $(x,y,z)$ in space. In general, we can describe a vector $v$ as an $n$-tuple of numbers: $v=(v_0,\ldots,v_{n-1})$. One way to store vectors in Python is by using *lists*: $v_i$ is stored as *v[i]*.

**Arrays** are a generalization of vectors where we can have multiple indices: $A_{i,j}$, $A_{i,j,k}$. In Python code this is represented as a nested list (see previous lecture), accessed as *A[i][j]*, *A[i][j][k]*.

Example: table of numbers, one index for the row, one for the column
$$
\left\lbrack\begin{array}{cccc}
0 & 12 & -1 & 5q
-1 & -1 & -1 & 0\cr
11 & 5 & 5 & -2
\end{array}\right\rbrack
\hspace{1cm}
A =
\left\lbrack\begin{array}{ccc}
A_{0,0} & \cdots &  A_{0,n-1}\cr
\vdots & \ddots &  \vdots\cr
A_{m-1,0} & \cdots & A_{m-1,n-1}
\end{array}\right\rbrack
$$
The number of indices in an array is the *rank* or *number of dimensions*. Using these terms, a vector can be described as a one-dimensional array, or rank 1 array.

In practice, we use [Numerical Python (*NumPy*)](http://www.numpy.org/) arrays instead of lists to represent mathematical arrays because it is **much** faster for large arrays.

Let's consider an example where we store $(x,y)$ points along a curve in Python lists and numpy arrays:

In [73]:
# Sample function
def f(x):
    return x**3

# Generate n points in [0,1]
n = 5
dx = 1.0/(n-1) # x spacing

X = [i*dx for i in range(n)] # Python lists
Y = [f(x) for x in X]

# Turn these Python lists into Numerical Python (NumPy) arrays:
import numpy as np
x2 = np.array(X)
y2 = np.array(Y)

Instead of first making lists with $x$ and $y = f (x)$ data, and then turning lists into arrays, we can make NumPy arrays
directly:

In [84]:
import numpy as np
n = 5                     # number of points
x2 = np.linspace(0, 1, n)    # generates n points between 0 and 1
y2 = np.zeros(n)             # n zeros (float data type by default)
for i in range(n):     
    y2[i] = f(x2[i])

List comprehensions create lists, not arrays, but we can do:

In [75]:
y2 = np.array([f(xi) for xi in x2]) # list -> array

### When and where to use NumPy arrays

* Python lists can hold any sequence of any Python objects, however, NumPy arrays can only hold objects of the same type.
* Arrays are most efficient when the elements are of basic number types (*float*, *int*, *complex*).
* In that case, arrays are stored efficiently in the computer's memory and we can compute very efficiently with the array elements.
* Mathematical operations on whole arrays can be done without loops in Python. For example,

In [85]:
import math

x = np.linspace(0, 2, 10001)
y = np.zeros(10001)
for i in range(len(x)):
    y[i] = math.sin(x[i])

can be coded as

In [78]:
import numpy as np
a= np.linspace(20,150,10)
b= np.zeros(10)
b = np.sin(a)v0=10;g=9.81
def y(t):
    return v0*t-0.5*g*t**2
t = np.linspace(0,(2*v0/g),0.1)
y = y(t)
print(t)
print(b)

SyntaxError: invalid syntax (<ipython-input-78-86453d1dd599>, line 4)

In [79]:
y = np.sin(x)

In the latter case the loop over all elements is now performed in a very efficient C function.

Operations on whole arrays, instead of using Python *for*-loops, is called vectorization and is a very **convenient**, **efficient** and therefore important programming technique to master.

Let's consider a simple vectorisation example: a loop to compute $x$ coordinates (*x2*) and $y=f(x)$ coordinates (*y2*) along a function curve:

In [80]:
x2 = np.linspace(0, 1, n)
y2 = np.zeros(n)
for i in range(n):
    y2[i] = f(x2[i])

This computation can be replaced by:

In [81]:
x2 = np.linspace(0, 1, n)
y2 = f(x2)

The advantage of this approach is:

* There is no need to allocate space for y2 (via the NumPy *zeros* function).
* There is no need for a loop.
* It is *much* faster.

## How vectorised functions work
Consider the function

In [82]:
def f(x):
    return x**3

$f(x)$ is intended for a number $x$, *i.e.* a *scalar*. So what happens when we call *f(x2)*, where *x2* is an NumPy array? **The function simply evaluates $x^3$ for an array x**. NumPy supports arithmetic operations on arrays, which correspond to the equivalent operations on each element, *e.g.*:

In [ ]:
x**3                # x[i]**3 forr all i
np.cos(x)              # cos(x[i]) for all i
x**3 + x*np.cos(x)     # x[i]**3 + x[i]*cos(x[i]) for all i
x/3*np.exp(-x*0.5)     # x[i]/3*exp(-x[i]*0.5) for all i 

In each of these cases a highly optimised C function is actually called to evaluate the expression. In this example, the *cos* function called for an *array* is imported from *numpy* rather than from the *math* module which only acts on scalars.

Notes:

* Functions that can operate on arrays are called **vectorized functions**.
* Vectorization is the process of turning a non-vectorized expression/algorithm into a vectorized expression/algorithm.
* Mathematical functions in Python automatically work for both scalar and array (vector) arguments, *i.e.* no vectorization is needed by the programmer.


### Watch out for references Vs. copies of arrays!
Consider this code:

In [ ]:
a=x
a[-1] = 42
print(x[-1])

Notice what happened here - we changed a value in *a* but the corresponding value in *x* was also changed! This is because *a* refers to the same array as *x*. If you really want a seperate copy of *x* then we have to make an explicit copy:

In [ ]:
a = x.copy()

## Exercise 2.8: Fill lists and arrays with function values

A function with many applications in science is defined as:

$$h(x) = \frac{1}{\sqrt{2\pi}}\exp(-0.5x^2)$$

* Implement the above formula as a Python function. Call the function *h* and it should take just one argument, *x*.
* Create a NumPy array (call it $x$) that has 9 uniformly spaced points in [−4, 4].
* Create a second NumPy array (call it $y$) with the function $h(x)$.

In [86]:
def h(x):
    from math import sqrt,exp,pi
    p = (1/sqrt(2*pi)) * exp(-0.5*x**2)
    return p
x = np.linspace(-4,4,9)
y = np.array([h(x[i]) for i in range(len(x))])
print(y)

[1.33830226e-04 4.43184841e-03 5.39909665e-02 2.41970725e-01
 3.98942280e-01 2.41970725e-01 5.39909665e-02 4.43184841e-03
 1.33830226e-04]


In [87]:
ok.grade('question-2_8')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 6
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 6}

## Generalised array indexing
We can select a slice of an array using *a[start:stop:inc]*, where the slice *start:stop:inc* implies a set of indices starting from *start*, up to *stop* in increments of *inc*. In fact, any integer list or array can be used to indicate a set of indices:

In [88]:
import numpy as np
a = np.linspace(1, 8, 8)
print(a)

[1. 2. 3. 4. 5. 6. 7. 8.]


In [89]:
a[[1,6,7]] = 10 # i.e. set the elements with indicies 1,6, and 7 in the list to 10.
print(a)

[ 1. 10.  3.  4.  5.  6. 10. 10.]


In [90]:
a[range(2,8,3)] = -2   # same as a[2:8:3] = -2
print(a)

[ 1. 10. -2.  4.  5. -2. 10. 10.]


Even boolean expressions can also be used to select part of an array(!)

In [91]:
print(a[a < 0]) # pick out all negative elements

[-2. -2.]


In [92]:
a[a < 0] = a.max() # if a[i]<0, set a[i]=10
print(a)

[ 1. 10. 10.  4.  5. 10. 10. 10.]


## Exercise 2.9: Explore array slicing

* Create a NumPy array called *w* with 31 uniformly spaced values ranging from 0 to 3.
* Using array slicing, create a NumPy array called *wbits* that starts from the $4^{th}$ element of *w*, excludes the final element of *w* and selects every $3^{rd}$ element.

In [93]:
import numpy as np
w = np.linspace(0,3,31)
wbits = w[range(3,30,3)]
print(w)
print(wbits)

[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7
 1.8 1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3. ]
[0.3 0.6 0.9 1.2 1.5 1.8 2.1 2.4 2.7]


In [94]:
ok.grade('question-2_9')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## 2D arrays
When we have a table of numbers,

$$
\left\lbrack\begin{array}{cccc}
0 & 12 & -1 & 5\cr
-1 & -1 & -1 & 0\cr
11 & 5 & 5 & -2
\end{array}\right\rbrack
$$

(*i.e.* a *matrix*) it is natural to use a two-dimensional array $A_{i,j}$ with one index for the rows and one for the columns:

$$
A = 
\left\lbrack\begin{array}{ccc}
A_{0,0} & \cdots &  A_{0,n-1}\cr
\vdots & \ddots &  \vdots\cr
A_{m-1,0} & \cdots & A_{m-1,n-1}
\end{array}\right\rbrack
$$

Let's recreate this array using NumPy:

In [95]:
import numpy as np
A = np.zeros((3,4))
A[0,0] = 0
A[1,0] = -1
A[2,0] = 11

A[0,1] = 12
A[1,1] = -1
A[2,1] = 5

A[0,2] = -1
A[1,2] = -1
A[2,2] = 5

# we can also use the same syntax that we used for nested lists

A[0][3] = 5
A[1][3] = 0
A[2][3] = -2

print(A)

[[ 0. 12. -1.  5.]
 [-1. -1. -1.  0.]
 [11.  5.  5. -2.]]


Next let's convert a nested list from a previous example into a 2D array:

In [96]:
Cdegrees = range(0, 101, 10)
Fdegrees = [9./5*C + 32 for C in Cdegrees]
table = [[C, F] for C, F in zip(Cdegrees, Fdegrees)]
print(table)

[[0, 32.0], [10, 50.0], [20, 68.0], [30, 86.0], [40, 104.0], [50, 122.0], [60, 140.0], [70, 158.0], [80, 176.0], [90, 194.0], [100, 212.0]]


In [97]:
# Convert this into a NumPy array:
table2 = np.array(table)
print(table2)

[[  0.  32.]
 [ 10.  50.]
 [ 20.  68.]
 [ 30.  86.]
 [ 40. 104.]
 [ 50. 122.]
 [ 60. 140.]
 [ 70. 158.]
 [ 80. 176.]
 [ 90. 194.]
 [100. 212.]]


To see the number of elements in each dimension:

In [98]:
print(table2.shape)

(11, 2)


*i.e.* 11 rows and 2 columns.

Let's write a loop over all array elements of A:

In [99]:
for i in range(table2.shape[0]):
    for j in range(table2.shape[1]):
        print('table2[%d,%d] = %g' % (i, j, table2[i,j]))

table2[0,0] = 0
table2[0,1] = 32
table2[1,0] = 10
table2[1,1] = 50
table2[2,0] = 20
table2[2,1] = 68
table2[3,0] = 30
table2[3,1] = 86
table2[4,0] = 40
table2[4,1] = 104
table2[5,0] = 50
table2[5,1] = 122
table2[6,0] = 60
table2[6,1] = 140
table2[7,0] = 70
table2[7,1] = 158
table2[8,0] = 80
table2[8,1] = 176
table2[9,0] = 90
table2[9,1] = 194
table2[10,0] = 100
table2[10,1] = 212


Alternatively:

In [100]:
for index_tuple, value in np.ndenumerate(table2):
    print('index %s has value %g' % (index_tuple, table2[index_tuple]))

index (0, 0) has value 0
index (0, 1) has value 32
index (1, 0) has value 10
index (1, 1) has value 50
index (2, 0) has value 20
index (2, 1) has value 68
index (3, 0) has value 30
index (3, 1) has value 86
index (4, 0) has value 40
index (4, 1) has value 104
index (5, 0) has value 50
index (5, 1) has value 122
index (6, 0) has value 60
index (6, 1) has value 140
index (7, 0) has value 70
index (7, 1) has value 158
index (8, 0) has value 80
index (8, 1) has value 176
index (9, 0) has value 90
index (9, 1) has value 194
index (10, 0) has value 100
index (10, 1) has value 212


We can also extract slices from multi-dimensional arrays as before. For example, extract the second column:

In [101]:
print(table2[:, 1]) # 2nd column (index 1)

[ 32.  50.  68.  86. 104. 122. 140. 158. 176. 194. 212.]


Play with this more complicated example:

In [102]:
t = np.linspace(1, 30, 30).reshape(5, 6)
print(t)

[[ 1.  2.  3.  4.  5.  6.]
 [ 7.  8.  9. 10. 11. 12.]
 [13. 14. 15. 16. 17. 18.]
 [19. 20. 21. 22. 23. 24.]
 [25. 26. 27. 28. 29. 30.]]


In [103]:
print(t[1:-1:2, 2:])

[[ 9. 10. 11. 12.]
 [21. 22. 23. 24.]]


## Exercise 2.10: Implement matrix-vector multiplication
A matrix $\mathbf{A}$ and a vector $\mathbf{b}$, represented in Python as a 2D array and a 1D array respectively, are given by:

$$
\mathbf{A} = \left\lbrack\begin{array}{ccc}
0 & 12 & -1\cr
-1 & -1 & -1\cr
11 & 5 & 5
\end{array}\right\rbrack
$$

$$
\mathbf{b} = \left\lbrack\begin{array}{c}
-2\cr
1\cr
7
\end{array}\right\rbrack
$$

Multiplying a matrix by a vector results in another vector $\mathbf{c}$, whose components are defined by the general rule

$$\mathbf{c}_i = \sum_j\mathbf{A}_{i,j}\mathbf{b}_j$$

* Define $\mathbf{A}$ and $\mathbf{b}$ as NumPy arrays
* Write a function called `multiply` that takes two arguments, a matrix and a vector in the form of NumPy arrays, and returns a NumPy array containing their product.
* Call this function on $\mathbf{A}$ and $\mathbf{b}$, and store the result in a variable $c$.

In [104]:
import numpy as np

def multiply(A,b):
    c = np.zeros(A.shape[0])
    for i in range(A.shape[0]):
        for j in range(b.shape[0]):
            c[i] += A[i,j]*b[j]
    return c
A = np.array([[0,12,-1],[-1,-1,-1],[11,5,5]])
b = np.array([-2,1,7])
c = multiply(A,b)
print(c)

[ 5. -6. 18.]


In [105]:
ok.grade('question-2_10')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 6
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 6}

## Exercise 2.11: Vectorised function

Let $A_{33}$ be the two-dimensional array
$$
\mathbf{A_{33}} = \left\lbrack\begin{array}{ccc}
0 & 12 & -1\cr
-1 & -1 & -1\cr
11 & 5 & 5
\end{array}\right\rbrack
$$

Implement and apply the function
$$
f(x) = x^3 + xe^x + 1
$$
to each element in $A_{33}$. Then calculate the result of the array expression ${A_{33}}^3 + A_{33}e^{A_{33}} + 1$, and demonstrate that the end result of the two methods are the same.

In [113]:
import numpy as np
def f_cubic(A):
    return A**3 + A*np.cos(A) + 1
a = [[0,12,-1],[-1,-1,-1],[11,5,5]]
A = np.array(a)
print(f_cubic(A))

a = [[0,12,-1],[-1,-1,-1],[11,5,5]]
A = np.array(a)
print(A**3 + A*np.cos(A) + 1)
        

[[ 1.00000000e+00  1.73912625e+03 -5.40302306e-01]
 [-5.40302306e-01 -5.40302306e-01 -5.40302306e-01]
 [ 1.33204868e+03  1.27418311e+02  1.27418311e+02]]
[[ 1.00000000e+00  1.73912625e+03 -5.40302306e-01]
 [-5.40302306e-01 -5.40302306e-01 -5.40302306e-01]
 [ 1.33204868e+03  1.27418311e+02  1.27418311e+02]]


In [114]:
ok.grade('question-2_11')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 3}

## Exercise 2.12: Implement matrix-matrix multiplication

Similarly to Exercise 4.3 (where a matrix is multiplied by a vector), the general rule for multiplying a $n \times m$ matrix $\mathbf{A}$ by a $m \times p$ matrix $\mathbf{B}$ results in a $n \times p$ matrix $\mathbf{C}$, whose components are defined by the general rule

$$\mathbf{C}_{i,j} = \sum^m_{k=1}\mathbf{A}_{i,k}\mathbf{B}_{k,j}$$

Again let $\mathbf{A}$ be the two-dimensional array

$$
\mathbf{A} = \left\lbrack\begin{array}{ccc}
0 & 12 & -1\cr
-1 & -1 & -1\cr
11 & 5 & 5
\end{array}\right\rbrack
$$

and let $\mathbf{B}$ be the two-dimensional array

$$
\mathbf{B} = \left\lbrack\begin{array}{ccc}
-2 & 1 & 7\cr
3 & 0 & 6\cr
2 & 3 & 5
\end{array}\right\rbrack
$$

Define `A` and `B` as NumPy arrays, and write a function `f_mult` which multiplies them together using the above rule to create a the NumPy array called `C`.

In [108]:
import numpy as np
def f_mult(A,B):
    if A.shape[1] != B.shape[0]:
        raise RunTimeError("A should be have the same number of columns as B has rows")
    C = np.zeros([A.shape[0],B.shape[1]])
    for i in range((A.shape[0])):
        for j in range((B.shape[0])):
            for k in range((B.shape[0])):
                C[i][j] += A[i][k]*B[k][j]
    return C
A = np.array([[0,12,-1],[-1,-1,-1],[11,5,5]])
B = np.array([[-2,1,7],[3,0,6],[2,3,5]])
C = f_mult(A,B)
print(C)


[[ 34.  -3.  67.]
 [ -3.  -4. -18.]
 [  3.  26. 132.]]


In [109]:
ok.grade('question-2_12')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 2}

## Exercise 2.13: 2D array slicing


* Create a 1D NumPy array called `odd` with all of the odd numbers from 1 to 55
* Create a 2D NumPy array called `odd_sq` with all of the odd numbers from 1 to 55 in a matrix with 4 rows and 7 columns
* Using array slicing, create a 2D NumPy array called, `odd_bits`, that starts from the $2^{nd}$ column of `odd_sq` and selects every other column, of only the $2^{nd}$ and $3^{rd}$ rows of `odd_sq`

In [110]:
import numpy as np
n = 55
odd_list = []
k = 1
while k <=n:
    if k%2 != 0:
        odd_list.append(k)
    k = k + 1
odd = np.array(odd_list)
odd_sq = odd.reshape(4,7)
odd_bits = odd_sq[1:3,1:-1:2]
print(odd_bits)

[[17 21 25]
 [31 35 39]]


In [111]:
ok.grade('question-2_13')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



{'failed': 0, 'locked': 0, 'passed': 3}

In [112]:
ok.score()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Scoring tests

---------------------------------------------------------------------
question 2.1
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
question 2.2
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
question 2.3
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
question 2.4
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
question 2.5
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
question 2.6
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed

---------------------------------------------------------------------
quest

{'Total': 13.0}